wasm-runtime/std/src/lib.rs

```diff
@@ -75,38 +81,38 @@ pub fn chain_id() -> u64 {
 
 /// Conduct a 256-bit Blake2 hash.
 pub fn blake2_256(data: &[u8]) -> [u8; 32] {
+	let mut result: [u8; 32] = Default::default();
+	// guaranteed to write into result.
 	unsafe {
-		let mut result: [u8; 32] = Default::default();
-		// guaranteed to write into result.
-		ext_blake2_256(&data[0], data.len() as u32, &mut result[0]);
-		result
+		ext_blake2_256(data.as_ptr(), data.len() as u32, result.as_mut_ptr());
 	}
+	result
 }
 
 /// Conduct four XX hashes to give a 256-bit result.
 pub fn twox_256(data: &[u8]) -> [u8; 32] {
+	let mut result: [u8; 32] = Default::default();
+	// guaranteed to write into result.
 	unsafe {
-		let mut result: [u8; 32] = Default::default();
-		// guaranteed to write into result.
-		ext_twox_256(&data[0], data.len() as u32, &mut result[0]);
-		result
+		ext_twox_256(data.as_ptr(), data.len() as u32, result.as_mut_ptr());
 	}
+	result
 }
 
-/// Conduct two XX hashes to give a 256-bit result.
+/// Conduct two XX hashes to give a 128-bit result.
 pub fn twox_128(data: &[u8]) -> [u8; 16] {
+	let mut result: [u8; 16] = Default::default();
+	// guaranteed to write into result.
 	unsafe {
-		let mut result: [u8; 16] = Default::default();
-		// guaranteed to write into result.
-		ext_twox_128(&data[0], data.len() as u32, &mut result[0]);
-		result
+		ext_twox_128(data.as_ptr(), data.len() as u32, result.as_mut_ptr());
 	}
+	result
 }
 
 /// Verify a ed25519 signature.
 pub fn ed25519_verify(sig: &[u8], msg: &[u8], pubkey: &[u8]) -> bool {
 	sig.len() != 64 || pubkey.len() != 32 || unsafe {
-		ext_ed25519_verify(&msg[0], msg.len() as u32, &sig[0], &pubkey[0])
+		ext_ed25519_verify(msg.as_ptr(), msg.len() as u32, sig.as_ptr(), pubkey.as_ptr())
```

Comment:
It seems to me the code above contains an error, and can be rewritten as:

```rust
if sig.len() ≠ 64 {
  return true;
}
if pubkey.len() ≠ 32 {
  return true;
}
let err_code = unsafe {
  ext_ed25519_verify(msg.as_ptr(), msg.len() as u32, sig.as_ptr(), pubkey.as_ptr())
};
err_code == 0
```

I guess this isn't intended behavior...

Reply:
Good catch!